contracts/near/src/lib.rs

```diff
@@ -0,0 +1,489 @@
+    /// upgrade has been allowed, calling `upgrade` with code that matches this hash will cause the
+    /// contract to upgrade itself.
+    pub codehash: [u8; 32],
+
+    /// Staleness threshold for rejecting price updates.
+    pub stale_threshold: u64,
+
+    /// Fee for updating price.
+    pub update_fee: Balance,
+}
+
+/// The `Action` acts as an enum of possible `Action`'s this contract can receive via VAA.
+#[derive(BorshDeserialize, BorshSerialize)]
+pub enum Action {
+    BatchAttest(Vec<u8>),
+    ContractUpgrade([u8; 32]),
+    SetDataSources(Vec<Source>),
+    SetGovernanceSource(Source),
+    SetStalePriceThreshold(u64),
+    SetUpdateFee(Balance),
+}
+
+#[near_bindgen]
+impl Pyth {
+    #[init]
+    #[allow(clippy::new_without_default)]
+    pub fn new(
+        wormhole: AccountId,
+        codehash: [u8; 32],
+        initial_source: Source,
+        gov_source: Source,
+        update_fee: Balance,
+        stale_threshold: u64,
+    ) -> Self {
+        // Add an initial Source so that the contract can be used.
+        let mut sources = UnorderedSet::new(StorageKeys::Source);
+        sources.insert(&initial_source);
+        Self {
+            prices: UnorderedMap::new(StorageKeys::Prices),
+            last_gov_seq: 0,
+            stale_threshold,
+            gov_source,
+            sources,
+            wormhole,
+            codehash,
+            update_fee,
+        }
+    }
+
+    #[init(ignore_state)]
+    pub fn migrate() -> Self {
+        let state: Self = env::state_read().expect("Failed to read state");
+        state
+    }
+
+    /// Instruction for processing VAA's relayed via Wormhole.
+    ///
+    /// Note that VAA verification requires calling Wormhole so processing of the VAA itself is
+    /// done in a callback handler, see `process_vaa_callback`.
+    #[payable]
+    #[handle_result]
+    pub fn update_price_feeds(&mut self, vaas: Vec<String>) -> Result<(), Error> {
+        for vaa_hex in vaas {
+            // We Verify the VAA is coming from a trusted source chain before attempting to verify
+            // VAA signatures. Avoids a cross-contract call early.
+            let vaa = hex::decode(&vaa_hex).map_err(|_| Error::InvalidHex)?;
+            let vaa = vaa_payload::from_slice_with_payload::<structs::Vaa<()>>(&vaa);
+            let vaa = vaa.map_err(|_| Error::InvalidVaa)?;
+            let (vaa, _rest) = vaa;
+
+            // Convert to local VAA type to catch APi changes.
+            let vaa = Vaa::from(vaa);
+
+            if self.sources.contains(&Source {
+                emitter:            vaa.emitter_address,
+                pyth_emitter_chain: vaa.emitter_chain,
+            }) {
+                return Err(Error::UnknownSource);
+            }
+
+            // Verify VAA and refund the caller in case of failure.
+            ext_wormhole::ext(env::current_account_id())
+                .with_static_gas(Gas(30_000_000_000_000))
+                .verify_vaa(vaa_hex.clone())
```

Comment:
I'm not sure I know WH nor near well here, but it seems that we parse the VAA and wormhole also does that. If there is a gas concern, is it possible to ask wormhole to return the parsed vaa?

Reply:
Unfortunately no. However the VAA parse here ignores the payload and is the serde deserializer is cheap.